### PR TITLE
feat: adds the ability to view draft posts in dev mode

### PR DESCRIPTION
**Pull request by SSmale**
Currently you need to make a post not in draft (set the frontmatter key `draft` to false to get it to render. This makes it easy to accidently commit a draft post with the value set to false when you want to see how it will render.

This change allows posts marked as draft to be seen when the builtin astro env variable is set to development (when it is run with `astro dev` or `npm start` for us). 

When the app is built (`astro build`) the env variable will be production and the usual filter is set.

https://docs.astro.build/en/guides/environment-variables/#default-environment-variables

### CODE REVIEW

**Comment by satnaing** [comment]
Great idea.
One thing about this is that some users might see this confusing even though titles of `draft` posts in DEV mode starts with the word `"DRAFT: "`. Better approach for this, in my opinion, is giving users some kind of toggle during DEV mode.
Maybe AstroPaper could make use of Astro v4 new [Dev Toolbar](https://docs.astro.build/en/reference/dev-toolbar-app-reference/) integration for this kind of feature. 💡
I haven't dig the API enough and thus not sure whether it is possible. 
But if  it is, it will be beneficial and more convenient for users.

**Comment by SSmale** [comment]
Thanks for the pointer, it is done. I have also [posted in the Astro Docs discussion](https://github.com/withastro/docs/discussions/6256) as the docs are hard to follow on the feature 

https://github.com/satnaing/astro-paper/assets/17761689/45595f28-0f62-4c73-aa1e-8f8a43f9e874

Not sure where you want to document the feature. it is off by default but can be set to true by having an env var of `DRAFT_MODE=true` either in an env file or as part of the run command `DRAFT_MODE=true npm start`

**Comment by SSmale** [comment]
It has been a while since I have used ts and things have moved on a little, also Astro is all new to me to I went with js to keep it simple. 

There are the canvas elements so you can build a control panel for more functions, which would be cool for the other parts. Just not sure what other things you'd want to toggle as it is only there in dev mode, I am not sure the use cases they imagined. 


I think it is already opt in by basically being hidden. It is going to be quite tightly coupled to AstroPaper so might not be much use standalone. Would you be able to use it on another theme? Do all Astro based blogs have the same front matter? I am one to keep things monolithic until there is a reason to break them out. I do get the icons and GitLab as blog posts but they are all in the one repo so you can check it out and have everything you need. 

Something that might need to go in before it is released is a way to suppress the logs. They do get noisy but are useful to see how it works.

**Comment by satnaing** [comment]
> It has been a while since I have used ts and things have moved on a little, also Astro is all new to me to I went with js to keep it simple.

Yep, understandable! 
I can co-author for that tho.

> There are the canvas elements so you can build a control panel for more functions, which would be cool for the other parts. Just not sure what other things you'd want to toggle as it is only there in dev mode, I am not sure the use cases they imagined.

I'm thinking something like social images and meta tags previews in dev mode. This could be beneficial and I've been debugging social images and meta tags using chrome extensions.
And maybe some config options inside the `src/config.ts` 🤷

> I think it is already opt in by basically being hidden. It is going to be quite tightly coupled to AstroPaper so might not be much use standalone. Would you be able to use it on another theme? Do all Astro based blogs have the same front matter? I am one to keep things monolithic until there is a reason to break them out. I do get the icons and GitLab as blog posts but they are all in the one repo so you can check it out and have everything you need.

This might be bias (maybe childish XD) but I wanted to publish a package since I've never done one. 😃
Also, there's a plan to add a CLI tool for AstroPaper, like scaffolding, adding extra integrations (features) etc.
Just like shadcn/ui, AstroPaper users will be able to add integrations on their own by simply installing that ingetration.
For example, currently, if a user want to add a `reading time` in their post, they have to follow along the blog post. But making a separate package would allow them to get this functionality by just running a command in their terminal.

File: src/toolbar/draftMode/draftModeIntergration.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+export default {
+  name: "Draft-Mode",
+  hooks: {
+    "astro:config:setup": ({ addDevToolbarApp }) => {
+      addDevToolbarApp("./src/toolbar/draftMode/draftModeClient.js");
+    },
+    "astro:server:setup": ({ server }) => {
+      draftModeServer(server);
+    },
```
Comment (satnaing) [inline src/toolbar/draftMode/draftModeIntergration.js:11]: Just want to know...
Is it necessary to add a draftModeServer file?
Is there any use case for that?
I just didn't think it was necessary. Plz, correct me if I'm wrong.

**Comment by SSmale** [comment]
Please feel free to update the code. 

Yeah that could be good, I think though if you want it to be an optional NPM package there will be some refactoring needed to decouple those parts, currently there are a lot of places that need changing, you would need some kind of hook system for the rpm package to override it. 

if you are looking to add these tools, it might be worth setting up an astropaper GitHub org to group them altogether.

File: src/toolbar/draftMode/draftModeIntergration.js (hunk 1 above)
Comment (SSmale) [reply src/toolbar/draftMode/draftModeIntergration.js:11]: If you haven't already, I would recommend running it up locally and watching the logs as you toggle it, it might be easier to understand than the explanation I am about to give.

From my understanding, Astro being server side rendered also happens in dev mode, you just get the recompile and browser reload. Therefore we need a way to inject the flag into the compile phase as the draft blog posts will never be parsed due to the filter. 

The Astro server part then has the listener for the web socket to update the flag and store the state. when the browser reloads, astro reruns the filter and now parses the extra files that are marked draft. 

The is the first time I have done anything with Astro or SSR and I am not too interested in how they work under the hood, I used the theme so I could just write my blog but there are some interesting bits to play with. So I just approached it as a simple problem rather than looking deeper

**Comment by satnaing** [comment]
@SSmale 

> if you are looking to add these tools, it might be worth setting up an astropaper GitHub org to group them altogether.

Yep. I'll consider doing so when I add more related integrations etc. If so, might need help since never done anything like that before.

For now, I'll merge it after I made a few minor changes.

File: src/pages/posts/[slug]/index.astro [hunk 2]
```diff
@@ -39,3 +40,4 @@ const pagination = getPagination({
 ---
 
 {post ? <PostDetails post={post} /> : <Posts {...pagination} />}
+@utils/postFilters
```
Comment (satnaing) [inline src/pages/posts/[slug]/index.astro:43]: What are these?
When I checkout this PR and run the project without these, it still working fine.

Comment (SSmale) [reply src/pages/posts/[slug]/index.astro:43]: In a lot of places around the code you have a filter for the array that checks if the draft flag is false. I moved that logic to one place so it was easy to change, keeping the code DRY

Comment (satnaing) [reply src/pages/posts/[slug]/index.astro:43]: Yes, I get that.
But what does line number 43 do?
In that file, `draftFilter` is already used in line number 15.
So, I don't think `@utils/postFilters` at line 43 is necessary.

Comment (SSmale) [reply src/pages/posts/[slug]/index.astro:43]: Ahh sorry, I replied to the comment without seeing it in context of the file. Not sure. That might be VSCode trying to be helpful

Comment (satnaing) [reply src/pages/posts/[slug]/index.astro]: Yes. 
When I rename some imports, VSCode tried to add those things.
I removed those in my new commit.

**Comment by satnaing** [comment]
Hello @SSmale 
I made a few changes and you can also give your opinion on that.
If we're satisfied with the result, we can proceed to resolve the conflicts.

**Comment by SSmale** [comment]
That all looks good to me.

**Comment by satnaing** [comment]
I noticed some bugs.

https://github.com/satnaing/astro-paper/assets/53733092/a7329c82-de0f-4ec6-8a6d-5bc138ec56a1

Also, in Firefox, there's a refresh bug which looks like an infinite loop in React.

https://github.com/satnaing/astro-paper/assets/53733092/fc87b851-0bf0-4fa9-ae89-52295627e923

**Comment by SSmale** [comment]
I had the infinite loop. It was due to the blocks not working. I think there was something about needing to restart the server after a change in writing the code as it didn't pick up one of the changes.

**Comment by satnaing** [comment]
@SSmale 
I think I found a way to solve this problem.
But it is a complete different approach.
I'll refactor, test and post it tmr since it's already midnight here.

**Comment by satnaing** [comment]
I've solved that issue using client side config and css only.
But I'm not satisfied with the result and I think it's still not the right approach.

Here's the result.

https://github.com/satnaing/astro-paper/assets/53733092/abf6f34c-457f-498a-a1df-13604f71a68c

**Comment by SSmale** [comment]
I’m not sure what that video is showing? 

Is it worth backing the changes out and testing my original code again, then slowly add the TS in, then move it to the package structure?

**Comment by SSmale** [comment]
Just tested it with my original PR code, Safari works fine, but get the reload loop on chrome and Firefox

**Comment by satnaing** [comment]
@SSmale 
The video is about toggling draft mode; and as you can see, the post `DRAFT:AstroPaper 4.0` is hidden and reappeared as we toggle the dev toolbar.
This is done by only using CSS.
But it has some trade-offs. For example, it only hides the draft posts visually. It doesn't actually remove from the DOM. Paginations can be incorrect and confusing by using this way.

**Comment by satnaing** [comment]
The best way to handle this, in my opinion, is sending events to `postFilter.ts` file when the toolbar is toggled. 
I can detect dev toolbar toggling inside `integration.ts` file.

```ts
      server.ws.on("astro-dev-toolbar:paper-dev-tool:toggled", data => {
          // can detect event inside this
      });
```

The thing is... that event cannot be detected inside the `postFilter` function.
I've tried `server.ws.send` and it's not working as expected.
I've also checked and tried [Vite HMR API](https://vitejs.dev/guide/api-hmr) but no progress. Maybe because I'm not familiar with that API 🤷.

**Comment by satnaing** [comment]
The main issue with this feature is that we can't directly control servercode using the Dev Toolbar API. While it enables communication between the client and server during rendering, it's not seamless with HMR.

This leads us to rely on JavaScript and DOM manipulation instead of server draft filtering, which affects DX. For example, enabling draft mode can sometimes cause 404 errors or infinite loops.

https://github.com/user-attachments/assets/cd95fd7d-d853-487d-84c2-e9f5022ccf7a

Given these challenges, I won’t merge this feature for now. However, I’ll definitely consider adding it once a better approach is found.

Thanks again for your contributions and discussions, @SSmale!

**Comment by SSmale** [comment]
Thanks for the update, is it worth going back to the original always on in dev mode in the interim?

**Comment by satnaing** [comment]
I don't quite get what you mean @SSmale

**Comment by SSmale** [comment]
The original PR was for it to show draft posts with a custom title prefix when the app was running in dev mode.

**Comment by satnaing** [comment]
The loop issue I mentioned still exists in the original PR.
You can test it by checking out [this commit](fe4229be932abc30b880213ae64eb0c650e9be8f).
If we open the two tabs at the same time, for some reason, the infinite loop begins.
So, for now, I'll focus on other issues and features.
However, thanks for this PR and your previous contributions. ✨

**Comment by SSmale** [comment]
> The loop issue I mentioned still exists in the original PR.
> 
> You can test it by checking out [this commit](fe4229be932abc30b880213ae64eb0c650e9be8f).

There were two approaches. [My original one here](https://github.com/satnaing/astro-paper/commit/b5dc56628ff56418bef82badde6f51c377e53620), where I am just changing the filter criteria based on the mode, rather than using the Astro dev tool bar. The tool bar approach was when the loop was introduced.
